### PR TITLE
Add functions for replace child and replace clause

### PR DESCRIPTION
**Pull request by jeffreylovitz**
Adds the new functions `cypher_astnode_set_child` and `cypher_ast_query_set_clause`, as well as fixing several bugs in the AST clone logic.

### CODE REVIEW

File: lib/src/ast.c [hunk 1]
```diff
@@ -563,6 +563,7 @@ cypher_astnode_t *cypher_ast_clone(const cypher_astnode_t *ast)
     {
         goto failure;
     }
+    free(children);
```
Comment (swilly22) [inline lib/src/ast.c:566]: was it leaking ? when do we call `cypher_ast_clone` ?

File: lib/src/ast_query.c [hunk 2]
```diff
@@ -174,6 +174,16 @@ const cypher_astnode_t *cypher_ast_query_get_clause(
 }
 
 
+void cypher_ast_query_set_clause(
+        cypher_astnode_t *astnode, cypher_astnode_t *clause, unsigned int index)
+{
+    REQUIRE_TYPE(astnode, CYPHER_AST_QUERY, NULL);
+    REQUIRE_TYPE(clause, CYPHER_AST_QUERY_CLAUSE, NULL);
+    struct query *node = container_of(astnode, struct query, _astnode);
+    node->clauses[index] = clause;
```
Comment (swilly22) [inline lib/src/ast_query.c:183]: What guarantees do we have that `node->clauses[index]` is allocated, in addition are we overwriting anything at `node->clauses[index]` ?

File: lib/src/ast_return.c [hunk 3]
```diff
@@ -115,9 +115,9 @@ cypher_astnode_t *clone(const cypher_astnode_t *self,
     }
     cypher_astnode_t *order_by = (node->order_by == NULL) ? NULL :
             children[child_index(self, node->order_by)];
-    cypher_astnode_t *skip = (node->skip != NULL) ? NULL :
+    cypher_astnode_t *skip = (node->skip == NULL) ? NULL :
             children[child_index(self, node->skip)];
-    cypher_astnode_t *limit = (node->limit != NULL) ? NULL :
+    cypher_astnode_t *limit = (node->limit == NULL) ? NULL :
```
Comment (swilly22) [inline lib/src/ast_return.c:120]: what are the side effects of these faulty conditions ?

File: lib/src/ast_with.c [hunk 4]
```diff
@@ -117,9 +117,9 @@ cypher_astnode_t *clone(const cypher_astnode_t *self,
     }
     cypher_astnode_t *order_by = (node->order_by == NULL) ? NULL :
             children[child_index(self, node->order_by)];
-    cypher_astnode_t *skip = (node->skip != NULL) ? NULL :
+    cypher_astnode_t *skip = (node->skip == NULL) ? NULL :
             children[child_index(self, node->skip)];
-    cypher_astnode_t *limit = (node->limit != NULL) ? NULL :
+    cypher_astnode_t *limit = (node->limit == NULL) ? NULL :
```
Comment (swilly22) [inline lib/src/ast_with.c:122]: what are the side effects of these faulty conditions ?

File: lib/src/astnode.h [hunk 5]
```diff
@@ -127,7 +127,7 @@ static inline int child_index(const cypher_astnode_t *node,
 {
     unsigned int i = 0;
     while (i < node->nchildren && node->children[i] != child)
```
Comment (swilly22) [inline lib/src/astnode.h:129]: Before the fix, isn't this an infinity loop ? for `child` which is different than `node->children[0]`

File: lib/src/cypher-parser.h.in [hunk 6]
```diff
@@ -1269,6 +1279,25 @@ const cypher_astnode_t *cypher_ast_query_get_clause(
...
+ *
+ * @param [node] The AST node.
+ * @param [clause] The AST clause node.
+ * @param [index] The index of the clause.
+ */
+void cypher_ast_query_set_clause(
+        cypher_astnode_t *astnode, cypher_astnode_t *clause,
+        unsigned int index);
+
+
```
Comment (swilly22) [inline lib/src/cypher-parser.h.in:1300]: Similar to `cypher_astnode_set_child` which seems generic why do we need this function,
Isn't it a specific version of the former?

**Review by swilly22 — CHANGES_REQUESTED** [review]
Looking good, I have a few questions

File: lib/src/ast.c (hunk 1 above)
Comment (jeffreylovitz) [reply lib/src/ast.c:566]: Previously, we never called `cypher_ast_clone` (and I'm guessing no one else did either, given how severe some of the bugs in its logic are). This was leaking before, yes.

File: lib/src/ast_query.c (hunk 2 above)
Comment (jeffreylovitz) [reply lib/src/ast_query.c:183]: No guarantees, though I can add an assertion.

We are overwriting data in our use case. Currently we are freeing the replaced node explicitly, though I can place that free here instead if preferred.

File: lib/src/astnode.h (hunk 5 above)
Comment (jeffreylovitz) [reply lib/src/astnode.h:129]: Correct! Only the `cypher_ast_clone` logic uses this function, which makes me think it was previously untested.

**Pull request merged**